### ISSUE TITLE
Incorrect work of GenericArguments#vector3d

### ISSUE THREAD

**Issue by Lignium** [SpongePowered/SpongeAPI#1917]
Hi, SpongePowered!
When calling command with the "#target" special word, Vector3d argument points to a side block.

For example:

```
H> - the player head
.  - air
S  - the side block (air)
B  - the target block

H>.....SB
```

**Comment by ST-DDT** [comment]
Have you verified this behavior for all directions (NESWUD)? IIRC there is/was something similar where the ray ended on the block surface and if you use the surface coordinate to get the block then you would end up with the wrong block

**Comment by Lignium** [comment]
I tested it again. A command receives a position of point at a side of block, but not a position of block. Quite unexpectedly... If it was conceived, then an argument should be made to transfer to command coordinates of blocks, that is, Vector3i.

**Comment by ST-DDT** [comment]
Could you post the code to help us to reproduce the error?

**Comment by Lignium** [comment]
Simple command with Vector3d argument. Cloud particles spawn on sides of blocks.
https://gist.github.com/Lignium/93346438e007c132f04b5a713d3b9194

**Comment by dualspiral** [comment]
This isn't a bug in the way you're thinking it is. The element is doing the right thing, returning a point on the boundary between two blocks. It's the assumption of which block the resulting co-ordinate will pick that's the issue.

> Have you verified this behavior for all directions (NESWUD)? IIRC there is/was something similar where the ray ended on the block surface and if you use the surface coordinate to get the block then you would end up with the wrong block

This is because [it's using a block ray to trace the line of sight to the absolute first point between air and block](https://github.com/SpongePowered/SpongeAPI/blob/bleeding/src/main/java/org/spongepowered/api/command/args/GenericArguments.java#L1584-L1588). The intersection point happens to be at `<something>.0`, which is the boundary between two blocks - so which one does it choose? Ultimately, it truncates the decimal - so hits on the high end of the co-ordinate of the block you're interested in will actually select the air block - this will happen if you are predominately looking west, north or down.

> I tested it again. A command receives a position of point at a side of block, but not a position of block. Quite unexpectedly... 

Maybe it should be clearer that this is the case, but it's using a block ray under the hood and it is choosing the **exact point the player is looking at**. The javadocs are lacking in this though, I would agree, hence the confusion.

> then an argument should be made to transfer to command coordinates of blocks, that is, Vector3i.

I agree that there is use for an argument that raytraces and selects a block, but in this case, I'd make it actually return a `BlockSnapshot`. Such an element could be supported directly in the API if that was a generally wanted feature.

You can, of course, create your own to do this, if you prefer.

**Comment by ST-DDT** [comment]
IMO we need both: targetBlock and targetBlockLocation. The first is used if you want to do something with the block (e.g. read the type), the second is used to mark a block (e.g. set spawnpoint or select edge of area selection)
Well in practise you could get the second value from the first.

**Comment by Lignium** [comment]
I think you need to leave everything as it is. You just need to create Vector3iCommandElement or something similar.

**Comment by ryantheleach** [comment]
I believe this is working as intended. 

@Lignium your suggestion for 'Vector3iCommandElement' Could work, but I'd prefer to see something that was explicitly a BlockPositionCommandElement, as it may give extra parsing/interpretation options, e.g. for #target you could select where a block would be placed (the air/liquid block) or the block itself.

**Comment by ryantheleach** [comment]
Closed in favor of #1935 in order to cleanup some of the confusion.
